test/test_session.py

```diff
@@ -171,6 +177,54 @@ def _test_ops(self, client, *ops):
                         "%s did not return implicit session to pool" % (f.__name__,),
                     )
 
+    def test_implicit_sessions_checkout(self):
+        # "To confirm that implicit sessions only allocate their server session after a
+        # successful connection checkout" test from Driver Sessions Spec.
+        succeeded = False
+        failures = 0
+        for _ in range(5):
+            listener = EventListener()
+            client = rs_or_single_client(
+                event_listeners=[listener], maxPoolSize=1, retryWrites=True
+            )
+            cursor = client.db.test.find({})
+            ops: List[Tuple[Callable, List[Any]]] = [
+                (client.db.test.find_one, [{"_id": 1}]),
+                (client.db.test.delete_one, [{}]),
+                (client.db.test.update_one, [{}, {"$set": {"x": 2}}]),
+                (client.db.test.bulk_write, [[UpdateOne({}, {"$set": {"x": 2}})]]),
+                (client.db.test.find_one_and_delete, [{}]),
+                (client.db.test.find_one_and_update, [{}, {"$set": {"x": 1}}]),
+                (client.db.test.find_one_and_replace, [{}, {}]),
+                (client.db.test.aggregate, [[{"$limit": 1}]]),
+                (client.db.test.find, [{}]),
+                (client.server_info, [{}]),
+                (client.admin.aggregate, [[{"$currentOp": {}}]]),
+                (cursor.distinct, ["_id"]),
+                (client.db.list_collections, []),
+            ]
+            threads = []
+            listener.results.clear()
+
+            for op, args in ops:
+                threads.append(ExceptionCatchingThread(target=op, args=args, name=op.__name__))
+                threads[-1].start()
+            self.assertEqual(len(threads), len(ops))
+            for thread in threads:
+                thread.join()
+                self.assertIsNone(thread.exc)
+            client.close()
+            lsid_set = set()
+            for i in listener.results["started"]:
+                if i.command.get("lsid"):
+                    lsid_set.add(i.command.get("lsid")["id"])
+            if len(lsid_set) == 1:
+                succeeded = True
+            else:
+                failures += 1
+        print(failures)
```

Comment:
Do we not need to assert that there were no failures?

Reply:
Nope, because we expect some failures, we just want at least one success.

Reply:
Gotcha